key-rotator/storage/key.go

```diff
@@ -0,0 +1,223 @@
+package storage
+
+import (
+	"bytes"
+	"context"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math/big"
+	"time"
+
+	"github.com/rs/zerolog/log"
+	k8smeta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8s "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	"github.com/abetterinternet/prio-server/key-rotator/key"
+)
+
+// Key represents a store of Prio keys, with functionality to read & write keys
+// from the store.
+type Key interface {
+	// PutBatchSigningKey writes the provided key as the batch signing key for
+	// the given (locality, ingestor) tuple, or returns an error on failure.
+	PutBatchSigningKey(ctx context.Context, locality, ingestor string, key key.Key) error
+
+	// PutPacketEncryptionKey writes the provided key as the packet encryption
+	// key for the given locality, or returns an error on failure.
+	PutPacketEncryptionKey(ctx context.Context, locality string, key key.Key) error
+
+	// GetBatchSigningKey gets the batch signing key for the given (locality,
+	// ingestor) pair, or returns an error on failure.
+	GetBatchSigningKey(ctx context.Context, locality, ingestor string) (key.Key, error)
+
+	// GetPacketEncryptionKey gets the packet encryption key for the given
+	// locality, or returns an error on failure.
+	GetPacketEncryptionKey(ctx context.Context, locality string) (key.Key, error)
+}
+
+// NewKey returns a new Key implementation, using the given Kubernetes secret
+// interface for backing storage.
+func NewKey(k8s k8s.SecretInterface, prioEnv string) (Key, error) { return k8sKey{k8s, prioEnv}, nil }
+
+type k8sKey struct {
+	k8s k8s.SecretInterface
+	env string // Prio environment name, e.g. "prod-us" or "prod-intl".
+}
+
+const (
+	liveVersionsSecretKey  = "secret_key"
+	keyVersionsSecretKey   = "key_versions"
+	secretKeyUnfilledValue = "not-a-real-key" // used in the secret_key secret key to denote no data
+)
+
+var _ Key = k8sKey{} // verify k8skey satisfies Key
+
+func (k k8sKey) PutBatchSigningKey(ctx context.Context, locality, ingestor string, key key.Key) error {
+	return k.putKey(ctx, k.batchSigningKeyName(locality, ingestor), key, serializeBatchSigningSecretKey)
+}
+
+func (k k8sKey) PutPacketEncryptionKey(ctx context.Context, locality string, key key.Key) error {
+	return k.putKey(ctx, k.packetEncryptionKeyName(locality), key, serializePacketEncryptionSecretKey)
+}
+
+func (k k8sKey) putKey(ctx context.Context, secretName string, key key.Key, serializeLiveVersions func(key.Key) ([]byte, error)) error {
+	log.Info().
+		Str("storage", "kubernetes").
+		Str("secret", secretName).
```

Comment:
It'll be obvious from the secret name, but we could also log a field for the secret "kind", i.e. batch signing vs packet encryption.